File: ProjectManagerUI/src/app/service/task.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http'
import { Observable, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { TaskModule } from '../model/task/task.module';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root'
})
export class TaskService {

  endpoint: string = `${environment.apiBaseUri}`;

  constructor(private http: HttpClient) {
  }

  
  // get all tasks
  getTasks(): Observable<TaskModule[]> {
    return this.http.get<TaskModule[]>(this.endpoint + '/api/task/gettasks')
    .pipe(
      catchError(e => this.handleError(e)));
  }

  getTasksByProjectId(projectId): Observable<TaskModule[]> {
    return this.http.get<TaskModule[]>(this.endpoint + '/api/task/gettasksbyprojectid/' + projectId)
    .pipe(
      catchError(e => this.handleError(e)));
  }

  getParentTasks(): Observable<TaskModule[]> {
    return this.http.get<TaskModule[]>(this.endpoint + '/api/task/getparenttasks')
    .pipe(
      catchError(e => this.handleError(e)));
  }

  // get task by id
  getTaskById(id: number): Observable<TaskModule> {
    return this.http.get<TaskModule>(this.endpoint + '/api/task/gettaskbyid/' + id)
    .pipe(
      catchError(e => this.handleError(e)));
  }

  // save Task
  saveTask(task: TaskModule): Observable<number> {
    return this.http.post<number>(this.endpoint + '/api/task/save', task)
    .pipe(
      catchError(e => this.handleError(e)));
  }

  // complete Task
  completeTask(id: number): Observable<boolean> {
    return this.http.post<boolean>(this.endpoint + '/api/task/complete', { TaskId: id })
    .pipe(
      catchError(e => this.handleError(e)));
  }
  
  // error handler
  handleError(error: any) {
    console.log(error);
    return throwError("Server error occured");
  }
}

File: ProjectManager.BusinessLib/Service/TaskService.cs
using System.Collections.Generic;

namespace ProjectManager.BusinessLib.Service
{
    using ProjectManager.BusinessLib.Interface;
    using ProjectManager.Model;
    using ProjectManager.DataAccessLib.Interface;

    public class TaskService : ITaskService
    {
        ITaskRepository _taskRepository;

        public TaskService(ITaskRepository taskRepository)
        {
            _taskRepository = taskRepository;
        }

        public List<TaskModel> GetTasks()
        {
            return _taskRepository.GetTasks();
        }

        public List<TaskModel> GetTasksByProject(int projectId)
        {
            return _taskRepository.GetTasksByProject(projectId);
        }

        public List<TaskModel> GetParentTasks()
        {
            return _taskRepository.GetParentTasks();
        }

        public TaskModel GetTaskById(int id)
        {
            return _taskRepository.GetTaskById(id);
        }

        public int Save(TaskModel model)
        {
            return _taskRepository.Save(model);
        } 

        public bool Delete(int id)
        {
            return _taskRepository.Delete(id);
        }

        public bool Complete(int id)
        {
            return _taskRepository.Complete(id);
        }
    }
}
